entities/cyclemanager/callbacks.go

```diff
@@ -0,0 +1,138 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2023 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package cyclemanager
+
+import (
+	"context"
+	"sync"
+)
+
+type UnregisterFunc func(ctx context.Context) error
+
+type callbacks interface {
+	register(cycleFunc CycleFunc) UnregisterFunc
+	execute(shouldBreak ShouldBreakFunc) bool
+}
+
+type callbackState struct {
+	cycleFunc  CycleFunc
+	runningCtx context.Context
+}
+
+type multiCallbacks struct {
+	lock    *sync.Mutex
+	counter uint32
+	states  map[uint32]*callbackState
+	keys    []uint32
+
+	canceledCtx context.Context
+}
+
+func newMultiCallbacks() callbacks {
+	canceledCtx, cancel := context.WithCancel(context.Background())
+	cancel()
+
+	return &multiCallbacks{
+		lock:        new(sync.Mutex),
+		counter:     0,
+		states:      map[uint32]*callbackState{},
+		keys:        []uint32{},
+		canceledCtx: canceledCtx,
```

Comment:
I used contexts to track if callback is being currently executed. Active context means callback is running, expired context means callback finished execution or was not yet started.

This particular canceled (expired) context was meant to be used as const context, for all newly added callbacks to mark them as not running.

I've changed the implementation to use nil to mark not yet run callbacks instead of expired ctx.